storybookjs/storybook
# Feature: Menu with core functionalities

#### Issues: #867 #314 [ui-#69](https://github.com/storybooks/storybook-ui/pull/69)

#### step towards: #124 #868 #862

At this moment we can only use keyboard shortcuts and query string to get core functionalities. In some cases, we'd like to do the same via mouse or touch screen. Especially necessary for access from **mobile devices**

### What I did

#### 1. Added menu duplicate keyboard shortcuts to the left (stories) panel

<details>
<summary>
expanded menu

</summary>

![image](https://cloud.githubusercontent.com/assets/14885189/25566043/92f160b2-2dfc-11e7-97ed-6e44c66cc77c.png)


</details>



<details>
<summary>
  This menu is collapsible (default state):

</summary>

![image](https://cloud.githubusercontent.com/assets/14885189/25566018/4cbae3fc-2dfc-11e7-98b4-d548367f9e4d.png)

</details>


<details>
<summary>
  This menu has buttons to navigate within stories / storyKinds

</summary>

![image](https://cloud.githubusercontent.com/assets/14885189/25566112/8921ac1c-2dfd-11e7-8897-b2974bc57341.png)

We'll need it in `Fullscreen` mode

</details>

<details>
<summary>
 Removed `⌘` button from the header of left panel

</summary>

![image](https://cloud.githubusercontent.com/assets/14885189/25566186/9b069202-2dfe-11e7-8cb6-210546d1ca3f.png)

It's still available from the core menu

</details>

#### 2. Added floating menu when left panel is hidden

<details>
<summary>
 Fullscreen mode

</summary>

![image](https://cloud.githubusercontent.com/assets/14885189/25566299/5ee8676c-2e00-11e7-981f-904911c11f6c.png)

So we can navigate via mouse/touchscreen or return to the normal view

</details>

<details>
<summary>
 Hidden left panel

</summary>

![image](https://cloud.githubusercontent.com/assets/14885189/25566330/db977d3e-2e00-11e7-9d6d-49f1300090a2.png)

</details>

<details>
<summary>
 Expanse direction

</summary>

![image](https://cloud.githubusercontent.com/assets/14885189/25566501/f6a0022e-2e03-11e7-93c6-f895fb6a657d.png)

Depending on its position menu expands to up or down direction

</details>

#### 3. Mantra modules API

- `jumpToKind` - new API to jump between storyKinds: [modules/api/actions](https://github.com/UsulPro/storybook/blob/core_func_menu/packages/storybook-ui/src/modules/api/actions/api.js#L95)

- `floatingBoxOptions` - settings of floating menu displaying. [Default](https://github.com/UsulPro/storybook/blob/core_func_menu/packages/storybook-ui/src/modules/ui/index.js#L13):

```js
floatingBoxOptions: { 
      showFloatingBox: true, 
      floatingBoxPosition: boxPositions.BOTTOM_LEFT, 
}
```

#### 4. Added tests for new components

### How to test

- Open [demo page](https://usulpro.github.io/storybook-ui/) to test usability. Try it with mobile devices.

- Clone repo
```shell
cd packages/storybook-ui
npm link
cd sample_repo
npm link @kadira/storybook-ui
npm start storybook
```

Try to use new menu to hide/show left/down panels. Go to `fullscreen`/`normal` mode. Navigate through the stories.

To change floating menu position set [default settings](https://github.com/UsulPro/storybook/blob/core_func_menu/packages/storybook-ui/src/modules/ui/index.js#L13) to one of [this](https://github.com/UsulPro/storybook/blob/core_func_menu/packages/storybook-ui/src/libs/menu_positions.js)

- Run tests

```shell
cd storybook
npm run test
```

All tests should pass

<details>
<summary> Details for review

</summary>

- New react component for menu is here: `packages/storybook-ui/src/modules/ui/components/core_menu/`

- It gets access to API via `react-komposer` here: `packages/storybook-ui/src/modules/ui/containers/core_menu.js`

-  new `jumpToKind` API is here: `packages/storybook-ui/src/modules/api/actions/api.js`

- all `svg` images for menu in the webpack bundled library: `packages/storybook-ui/src/modules/ui/components/assets/svg_package.js`. we can discuss this solution

</details>
